Source/Particles/Collision/BinaryCollision/PhotonPhoton/SinglePhotonPhotonCollisionEvent.H

```diff
@@ -0,0 +1,183 @@
+ * @param[in] u2x,u2y,u2z momenta of the second colliding particle
+ * @param[in] w1,w2 effective weight of the colliding particles
+ * @param[in] dt is the time step length between two collision calls.
+ * @param[in] dV is the volume of the corresponding cell.
+ * @param[in] pair_index is the index of the colliding pair
+ * @param[out] p_mask is a mask that will be set to true if pair (electron-positron) production
+ * occurs for that pair of macrophotons
+ * @param[out] p_pair_reaction_weight stores the weight of the product particles
+ * @param[in] event_multiplier factor used to increase the number of pair-production events by
+ * decreasing the weight of the produced particles
+ * @param[in] multiplier_ratio factor used to take into account unsampled pairs (i.e. the fact
+ * that a particle only collides with one or few particles of the other species)
+ * @param[in] probability_threshold probability threshold above which we decrease the event
+ * multiplier
+ * @param[in] probability_target_value if the probability threshold is exceeded, this is used
+ * to determine by how much the event multiplier is reduced
+ * @param[in] engine the random engine.
+ */
+template <typename index_type>
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void SinglePhotonPhotonCollisionEvent (const amrex::ParticleReal& u1x, const amrex::ParticleReal& u1y,
+                               const amrex::ParticleReal& u1z, const amrex::ParticleReal& u2x,
+                               const amrex::ParticleReal& u2y, const amrex::ParticleReal& u2z,
+                               amrex::ParticleReal w1, amrex::ParticleReal w2,
+                               const amrex::Real& dt, const amrex::ParticleReal& dV, const int& pair_index,
+                               index_type* AMREX_RESTRICT p_mask,
+                               amrex::ParticleReal* AMREX_RESTRICT p_pair_reaction_weight,
+                               const amrex::ParticleReal& event_multiplier,
+                               const int& multiplier_ratio,
+                               const amrex::ParticleReal& probability_threshold,
+                               const amrex::ParticleReal& probability_target_value,
+                               const amrex::RandomEngine& engine)
+{
+    // General notations in this function:
+    //     x_sq denotes the square of x
+    //     x_star denotes the value of x in the center of mass frame
+
+    using namespace amrex::literals;
+
+    const amrex::ParticleReal w_min = amrex::min(w1, w2);
+    const amrex::ParticleReal w_max = amrex::max(w1, w2);
+
+    constexpr auto one_pr = amrex::ParticleReal(1.);
+    constexpr auto two_pr = amrex::ParticleReal(2.);
+    constexpr auto one_half_pr = amrex::ParticleReal(1./2.);
+    constexpr amrex::ParticleReal c = PhysConst::c;
+    constexpr amrex::ParticleReal c_sq = c*c;
+    constexpr amrex::ParticleReal me = PhysConst::m_e;
+    auto constexpr pow2 = [](double const x) { return x*x; };
```

Comment:
I would suggest using `auto` here
```suggestion
    auto constexpr pow2 = [](auto const x) { return x*x; };
```